Userland/Libraries/LibGUI/FileSystemModel.cpp

```diff
@@ -125,21 +127,18 @@ void FileSystemModel::Node::traverse_if_needed()
     children.append(move(directory_children));
     children.append(move(file_children));
 
-    if (!m_file_watcher) {
-
-        // We are not already watching this file, create a new watcher
-        auto watcher_or_error = Core::FileWatcher::watch(full_path);
-
-        // Note : the watcher may not be created (e.g. we do not have access rights.) This is expected, just don't watch if that's the case.
-        if (!watcher_or_error.is_error()) {
-            m_file_watcher = watcher_or_error.release_value();
-            m_file_watcher->on_change = [this](auto) {
-                has_traversed = false;
-                mode = 0;
-                children.clear();
-                reify_if_needed();
-                m_model.did_update();
-            };
+    if (!m_model.m_file_watcher->is_watching(full_path)) {
+        // We are not already watching this file, watch it
+        auto result = m_model.m_file_watcher->add_watch(full_path,
+            InodeWatcherEvent::Type::MetadataModified
+                | InodeWatcherEvent::Type::ChildCreated
+                | InodeWatcherEvent::Type::ChildDeleted
+                | InodeWatcherEvent::Type::Deleted);
+
+        if (result.is_error()) {
+            dbgln("Couldn't watch '{}': {}", full_path, result.error());
+        } else if (result.value() == false) {
+            dbgln("Couldn't watch '{}', probably already watching", full_path);
```

Comment:
Can we report the actual error here instead?

Reply:
The only two errors possible when `result.value() == false` are the path being invalid and us already watching the path. I could reword it to `already watching or the path was invalid` instead. I feel like adding yet another explicit error enum wouldn't really add much but would make the checks more annoying.